### PR TITLE
feat: Array element extraction

### LINKED ISSUE #898: Support array extraction
### What is the problem the feature request solves?

Support for extracting elements from an array. Spark has two expressions for doing this: `GerArrayItem` and `ElementAt`. These have a lot in common, with only minor differences for zero or one based, negative indices, and a default value, so they can probably be combined into a single DataFusion expression. 

### Describe the potential solution

Currently DataFusion has an `ArrayElement` expression which acts like `ElementAt`, but without the fail on error ANSI support and default value support, which is only needed for the `SplitPart` Spark expression via runtime replacement. It makes more sense to implement a custom function to support these behaviors, as well as also handle the `GetArrayItem` functionality that is zero based and doesn't support negative indices.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kimahriman**
## Which issue does this PR close?

Closes #898 

## Rationale for this change

More array/list support. Implemented a custom PhysicalExpr rather than using DataFusion's `ArrayElement` to support ANSI fail on error, default value, and GetArrayItem's zero based indexing.

## What changes are included in this PR?

Support for array extraction via the Spark `GetArrayItem` and `ElementAt` expressions.

## How are these changes tested?

New UT

### CODE REVIEW

File: native/spark-expr/src/list.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+    fn children(&self) -> Vec<&Arc<dyn PhysicalExpr>> {
+        vec![&self.child, &self.ordinal]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn PhysicalExpr>>,
+    ) -> datafusion_common::Result<Arc<dyn PhysicalExpr>> {
+        Ok(Arc::new(ListExtract::new(
```
Comment (andygrove) [inline native/spark-expr/src/list.rs:157]: nit: perhaps add an assertion that `children` is of the expected length

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @Kimahriman

File: native/spark-expr/src/list.rs (hunk 1 above)
Comment (Kimahriman) [reply native/spark-expr/src/list.rs:157]: Sure. Out of curiosity is this part of physical expressions (`with_new_children`) actually used in how Comet is using DataFusion?

Comment (andygrove) [reply native/spark-expr/src/list.rs]: It will be once https://github.com/apache/datafusion-comet/pull/907 is merged

**Pull request merged**